src/app/cluster/cluster-details/constraints/component.ts

```diff
@@ -0,0 +1,182 @@
+import {Cluster} from '@shared/entity/cluster';
+import {Constraint, Kind, Violation} from '@shared/entity/opa';
+import {UserSettings} from '@shared/entity/settings';
+import * as _ from 'lodash';
+import {Subject} from 'rxjs';
+import {take, takeUntil, filter, switchMap} from 'rxjs/operators';
+import {Mode, ConstraintDialog} from './constraint-dialog/component';
+
+@Component({
+  selector: 'km-constraint-list',
+  templateUrl: './template.html',
+  styleUrls: ['./style.scss'],
+})
+export class ConstraintsComponent implements OnInit, OnChanges, OnDestroy {
+  @Input() cluster: Cluster;
+  @Input() projectID: string;
+  @Input() isClusterRunning: boolean;
+  @Input() constraints: Constraint[] = [];
+  settings: UserSettings;
+  dataSource = new MatTableDataSource<Constraint>();
+  displayedColumns: string[] = ['constraintName', 'constraintTemplate', 'match', 'violations', 'actions'];
+  toggledColumns: string[] = ['violationDetails'];
+  isShowDetails = [];
+  @ViewChild(MatSort, {static: true}) sort: MatSort;
+  paginator: MatPaginator;
+  @ViewChild(MatPaginator)
+  set matPaginator(mp: MatPaginator) {
+    this.paginator = mp;
+    if (this.paginator && this.settings) {
+      this.paginator.pageSize = this.settings.itemsPerPage;
+    }
+    this.dataSource.paginator = this.paginator;
+    this._cdr.detectChanges();
+  }
+
+  private readonly _unsubscribe = new Subject<void>();
+
+  constructor(
+    private readonly _opaService: OPAService,
+    private readonly _userService: UserService,
+    private readonly _cdr: ChangeDetectorRef,
+    private readonly _matDialog: MatDialog,
+    private readonly _notificationService: NotificationService
+  ) {}
+
+  ngOnInit(): void {
+    this.dataSource.data = this.constraints;
+    this.dataSource.sort = this.sort;
+    this.dataSource.paginator = this.paginator;
+    this.sort.active = 'constraintName';
+    this.sort.direction = 'asc';
+
+    this._userService.currentUserSettings
+      .pipe(takeUntil(this._unsubscribe))
+      .subscribe(settings => (this.settings = settings));
+  }
+
+  ngOnChanges(changes: SimpleChanges): void {
+    if (changes.constraints) {
+      this.dataSource.data = this.constraints;
+    }
+  }
+
+  ngOnDestroy(): void {
+    this._unsubscribe.next();
+    this._unsubscribe.complete();
+  }
+
+  isPaginatorVisible(): boolean {
+    return (
+      this.constraints &&
+      this.constraints.length > 0 &&
+      this.paginator &&
+      this.constraints.length > this.paginator.pageSize
+    );
+  }
+
+  isLoadingData(data: Constraint[]): boolean {
+    return _.isEmpty(data) && !this.isClusterRunning;
+  }
+
+  hasNoData(data: Constraint[]): boolean {
+    return _.isEmpty(data) && this.isClusterRunning;
+  }
+
+  toggleDetails(element: Constraint): void {
+    this.isShowDetails[element.name] = !this.isShowDetails[element.name];
+  }
+
+  displayKindNames(element: Kind[]): string {
+    const kinds = [];
+    element.map(x => kinds.push(x.kinds));
```

Comment:
`map` is maps values. It does not work the same as `forEach`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map